pkg/chunk/aws_storage_client.go

```diff
@@ -331,28 +340,232 @@ func (a dynamoDBRequestAdapter) HasNextPage() bool {
+	}
+
+	return append(dynamoDBChunks, s3Chunks...), nil
+}
+
+func (a awsStorageClient) getS3Chunks(ctx context.Context, chunks []Chunk) ([]Chunk, error) {
+	incomingChunks := make(chan Chunk)
+	incomingErrors := make(chan error)
+	for _, chunk := range chunks {
+		go func(chunk Chunk) {
+			chunk, err := a.getS3Chunk(ctx, chunk)
+			if err != nil {
+				incomingErrors <- err
+				return
+			}
+			incomingChunks <- chunk
+		}(chunk)
+	}
+
+	result := []Chunk{}
+	errors := []error{}
+	for i := 0; i < len(chunks); i++ {
+		select {
+		case chunk := <-incomingChunks:
+			result = append(result, chunk)
+		case err := <-incomingErrors:
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
+		return nil, errors[0]
+	}
+	return result, nil
+}
+
+func (a awsStorageClient) getS3Chunk(ctx context.Context, chunk Chunk) (Chunk, error) {
 	var resp *s3.GetObjectOutput
 	err := instrument.TimeRequestHistogram(ctx, "S3.GetObject", s3RequestDuration, func(ctx context.Context) error {
 		var err error
 		resp, err = a.S3.GetObjectWithContext(ctx, &s3.GetObjectInput{
 			Bucket: aws.String(a.bucketName),
-			Key:    aws.String(key),
+			Key:    aws.String(chunk.externalKey()),
 		})
 		return err
 	})
 	if err != nil {
-		return nil, err
+		return Chunk{}, err
 	}
 	defer resp.Body.Close()
 	buf, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return nil, err
+		return Chunk{}, err
+	}
+	if err := chunk.decode(buf); err != nil {
+		return Chunk{}, err
 	}
-	return buf, nil
+	return chunk, nil
 }
 
-func (a awsStorageClient) PutChunk(ctx context.Context, key string, buf []byte) error {
+func (a awsStorageClient) getDynamoDBChunks(ctx context.Context, chunks []Chunk) ([]Chunk, error) {
+	outstanding := dynamoDBReadRequest{}
+	chunksByKey := map[string]Chunk{}
+	for _, chunk := range chunks {
+		key := chunk.externalKey()
+		chunksByKey[key] = chunk
+		tableName := a.chunkTableFor(chunk.From)
+		outstanding.Add(tableName, key, nil)
+	}
+
+	result := []Chunk{}
+	unprocessed := dynamoDBReadRequest{}
+	backoff, numRetries := minBackoff, 0
+	for outstanding.Len()+unprocessed.Len() > 0 {
+		requests := dynamoDBReadRequest{}
+		requests.takeReqs(unprocessed, dynamoDBMaxReadBatchSize)
+		requests.takeReqs(outstanding, dynamoDBMaxReadBatchSize)
+
+		var response *dynamodb.BatchGetItemOutput
+		err := instrument.TimeRequestHistogram(ctx, "DynamoDB.BatchGetItemPages", dynamoRequestDuration, func(ctx context.Context) error {
+			var err error
+			response, err = a.DynamoDB.BatchGetItemWithContext(ctx, &dynamodb.BatchGetItemInput{
+				RequestItems:           outstanding,
+				ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+			})
+			return err
+		})
+
+		for _, cc := range response.ConsumedCapacity {
+			dynamoConsumedCapacity.WithLabelValues("DynamoDB.BatchGetItemPages").
+				Add(float64(*cc.CapacityUnits))
+		}
+
+		if err != nil {
+			for tableName := range outstanding {
+				recordDynamoError(tableName, err)
+			}
+		}
```

Comment:
I wonder if it would make sense for everything from `var response *dynamodb.BatchGetItemOutput` (line 443) to this point  and making it a method (e.g. `queryDynamo`) on `dynamoDBReadRequest`.